### ISSUE TITLE
Allow full logging override with custom configuration

### ISSUE THREAD

**Issue by saulbein** [PrefectHQ/prefect#3712]
## Current behavior
The current logging settings in prefect allow only very coarse-grained control over logging - setting the global logging level (but what if I want to log my own code in debug level, but avoid having everything else spam me with debug messages?), or adding additional loggers (which partially solves the problem, but then a list of loggers must be maintained, which can quickly become a  hassle, and I would prefer to have the opposite policy of allowing logs until proven too spammy). Adding the root logger `''` to additional loggers _almost works_, but causes logs to be duplicated, as it then also reconfigures prefect's own logger as well.

## Proposed behavior
It would be much more preferable to be able to use the dictionary-style logging configuration supported by the standard python library, while keeping the current logging handlers as a default option. The user would supply either a logging configuration through the config file itself, or perhaps an importable variable string to use, which, since this is a plain python dictionary, could be easily serialized.

The main question here is whether it would be better to extend the logging configuration to add Prefect Server / Cloud logging handlers automatically, or simply provide a handler that could be used by the user (https://docs.python.org/3/howto/logging.html#handlers).

The default settings could then be combined into a default dictionary config by prefect, thus also providing an example configuration for users.

Another question is whether a simple logging handler can deal with the `stdout` redirection that prefect does. My personal opinion is that using `print` statements for logging should at least be discouraged, and most libraries I've used avoid printing, with some exceptions (looking at you, AWS SageMaker library).

## Example
This is essentially how I imagine the configuration could look like:
```
LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "typical": {
            # custom logging format
            "format": "[%(asctime)s.%(msecs)03d] [%(levelname)s] [%(threadName)s] [%(name)s:%(lineno)d] %(message)s",
            "datefmt": "%Y-%m-%d %H:%M:%S",
        },
    },
    "handlers": {
        # more handlers could be added here if I need to log to an external system, for example Graylog
        "console": {"class": "logging.StreamHandler", "formatter": "typical"},
        "prefect": {"class": "prefect.CloudHandler", "formatter": "typical"},
    },
    "loggers": {
        # this is sufficient for most debugging purposes
        "prefect": {"level": "INFO"},
        # I want granular logs for my own code
        "my_own_liibrary": {"level": "DEBUG"},
        # log SQL statements, but don't send them to prefect
        "sqlalchemy.engine": {"level": "INFO", handlers: ["console"], propagate: False},
    },
    "root": {
        # log everything by default, makes it easier to find which libraries can log useful info
        # log level here is global (unless propagation is off), could be modified via the log level variable
        "handlers": ["console", "prefect"], "level": "NOTSET"
    },
}
```
This how I usually set it in projects via code, but instead the structure could be replicated in TOML.

**Comment by jcrist** [comment]
The logging config provided by Python is great for configuring the logging of multiple libraries - the logging docs advise that libraries don't add any log handlers themselves, leaving that on the end user (https://docs.python.org/3/howto/logging.html#configuring-logging-for-a-library):

> Note:
> It is strongly advised that you do not add any handlers other than NullHandler to your library’s loggers. This is because the configuration of handlers is the prerogative of the application developer who uses your library. The application developer knows their target audience and what handlers are most appropriate for their application: if you add handlers ‘under the hood’, you might well interfere with their ability to carry out unit tests and deliver logs which suit their requirements.

Unfortunately this doesn't mesh as well with tools that are both libraries *and* applications (as prefect is - dask also runs into this problem).

- When used as a library, it'd be best for Prefect not to configure any logging handlers and rely on the user to enable loggers as needed. However, we also want an easy way for users to configure logging (usually just log levels), without having them have to go read the logging docs.

- When used as an application, we'd like the logging configuration to be managed by prefect, since we need to enable/disable cloud logs as needed, migrate logging configuration throughout a dask cluster, etc...

Supporting both expert users who want full customization of all logging and also new users who just want things to work is hard. To give you the flexibility you want, but still keep the simplicity most of our users want I propose the following:

- Add a `logging.config` field (or some better name) to our configuration that takes a full [dict config style configuration](https://docs.python.org/3/library/logging.config.html#dictionary-schema-details). This will be applied *after* the existing behavior, and any loggers that are specified here will overwrite the previous configuration.
- Change how `extra_loggers` are configured to better handle adding the root logger (we can special case this to do the right thing). I still don't think you should do this, but we can support it nicer.
- Maybe we move all prefect logging under a single `PrefectHandler` (rather than both a `CloudHandler` and `StreamHandler`) that handles both the local streaming logs and the sending to cloud. This makes it easier for users to register custom loggers with our handlers, since they'll only need to specify one. We might also want to add a default `prefect` handler in the `handlers` field of the dict config (if not present) so users can reference an implicit `prefect` handler as needed in the config.

**Comment by saulbein** [comment]
> Add a `logging.config` field (or some better name) to our configuration that takes a full dict config style configuration. This will be applied after the existing behavior, and any loggers that are specified here will overwrite the previous configuration.

Not sure if you're aware, but dict config has a flag whether to disable existing loggers, although it can cause some unusual issues and I prefer not to use it. I think the best way here is to have a base config (this is what log level, format and extra loggers would modify), use it when no dict config is specified. When dict config is specified, I think it should be a full, transparent and total override, preferably with no special or implicit prefect behavior.

> Change how `extra_loggers` are configured to better handle adding the root logger (we can special case this to do the right thing). I still don't think you should do this, but we can support it nicer.

I used this one only *because* there is no dict config available, I don't think adding the root logger here needs to be supported, it's more of a hack-around.

> Maybe we move all prefect logging under a single `PrefectHandler` (rather than both a `CloudHandler` and `StreamHandler`)

In the example above, I used the `StreamHandler` as logging to the local stdout, rather than somewhere managed by prefect. I think it's better to be explicit - logs go to the prefect cloud *and* the stream, which for me is useful, as stdout is then automatically sent to CloudWatch. If a user is using dict config, they probably know what they want, and I think hidden implicit stuff here would just cause unnecessary confusion.
